pkg/runtime/grpc/grpc-runtime.go

```diff
@@ -0,0 +1,418 @@
+		return nil, fmt.Errorf("unmarshaling catalog: %w", err)
+	}
+
+	return catalog, nil
+}
+
+func (r *Runtime) storeCatalog(catalog *runtime.Catalog) error {
+	configFile, err := r.getCatalogFilename()
+	if err != nil {
+		return fmt.Errorf("get catalog filename: %w", err)
+	}
+	err = os.MkdirAll(filepath.Dir(configFile), 0o750)
+	if err != nil && errors.Is(err, os.ErrExist) {
+		return fmt.Errorf("create config dir: %w", err)
+	}
+	catalogJSON, err := json.Marshal(catalog)
+	if err != nil {
+		return fmt.Errorf("marshaling catalog JSON: %w", err)
+	}
+	err = os.WriteFile(configFile, catalogJSON, 0o644)
+	if err != nil {
+		return fmt.Errorf("write catalog file: %w", err)
+	}
+	return nil
+}
+
+func (r *Runtime) Init(runtimeGlobalParams *params.Params) error {
+	catalog, err := r.loadLocalGadgetCatalog()
+	if err == nil {
+		r.catalog = catalog
+		return nil
+	}
+
+	// fallback to embedded catalog
+	catalog = &runtime.Catalog{}
+	err = json.Unmarshal(embeddedCatalog, &catalog)
+	if err == nil {
+		r.catalog = catalog
+		return nil
+	}
+
+	// Try remote
+	catalog, err = r.loadRemoteGadgetCatalog()
+	if err != nil {
+		return fmt.Errorf("could not get catalog: %w", err)
+	}
+	err = r.storeCatalog(catalog)
+	if err != nil {
+		log.Warnf("could not store catalog: %v", err)
+	}
+	r.catalog = catalog
+	return nil
+}
+
+func (r *Runtime) Close() error {
+	return nil
+}
+
+func (r *Runtime) ParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:         ParamNode,
+			Description: "Comma-separated list of nodes to run the gadget on",
+		},
+	}
+}
+
+func (r *Runtime) GlobalParamDescs() params.ParamDescs {
+	return params.ParamDescs{
+		{
+			Key:            ParamConnectionMethod,
+			Description:    "Method that should be used to connect to the Inspektor Gadget nodes",
+			PossibleValues: []string{"kubeapi-server-per-node", "grpc-direct"},
+			DefaultValue:   "kubeapi-server-per-node",
+		},
+	}
+}
+
+func (r *Runtime) getGadgetPods(ctx context.Context, nodes []string) ([]v1.Pod, error) {
+	config, err := utils.KubernetesConfigFlags.ToRESTConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to creating RESTConfig: %w", err)
+	}
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set up trace client: %w", err)
+	}
+
+	opts := metav1.ListOptions{LabelSelector: "k8s-app=gadget"}
+	pods, err := client.CoreV1().Pods("gadget").List(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("getting pods: %w", err)
+	}
+
+	if len(pods.Items) == 0 {
+		return nil, fmt.Errorf("no gadget pods found. Is Inspektor Gadget deployed?")
+	}
+
+	res := pods.Items
+
+	if len(nodes) > 0 {
+		res = make([]v1.Pod, 0, len(pods.Items))
+
+		// Filter nodes
+		for _, pod := range pods.Items {
+			found := false
+			for _, node := range nodes {
+				if node == pod.Spec.NodeName {
+					found = true
+					break
+				}
+			}
+			if !found {
+				continue
+			}
+			res = append(res, pod)
+		}
+	}
```

Comment:
What about https://github.com/inspektor-gadget/inspektor-gadget/commit/bdab0a28189113ce85f97c111ecd75917273bc65 to printing a warning in case the node doesn't exist?